Possibility to parse SBOM string

trueserjo opened the pull request:
It might be useful when you receive a SBOM file as an InMemoryUploadedFile. You probably want to parse it without saving to disk.

anthonyharrison reviewed (changes requested):
Seems a reasonable idea. I wonder if we should change the parse function (and the equivalent functions for spdx and cyclonedx) to have either then filename or the file_str. I think this can be achieved by using a decorator to force this mutual exclusion.

The documentation (README) needs to be updated to reflect this new capability.

trueserjo commented:
I updated README.
The parser part is a bit more tricky. In the main parser (parse_file) filename and file_str are not mutually exclusive, because you determine what parser to use based on the filename. So it is needed either way. In the cyclone and spdx parsers these parameters are indeed mutually exclusive. Filename, if provided in addition to file_str, doesn't break anything, but I agree that it is better to make things more straightforward. I refactored parameters in a way that they are both needed now. I think it introduces less new code comparing to the decorator.

anthonyharrison reviewed (changes requested):
Thanks but I wonder if it would be better to have two separate functions. 

One to parse a filename as current, and one to parse as string. They can call the same underlying functions but it makes it cleaner to the use. This is the same way that the Python json library works with json.load (for filename) and json.loads (for string).

So I would suggest creating a new function called parse_string as this is then more explicit what is being done.

trueserjo commented:
Agree. Done.

anthonyharrison commented on README.md:
```diff
@@ -94,8 +94,13 @@ Errors in the parsing process will result in the exception SBOMParserException b
+
+If an error occurs during the parsing of the file, a SBOMParserException exception is raised.
+
+parse_string(filename, file_str)
```
Surely there is only a single parameter - file_str or more correctly sbom_string

trueserjo commented on lib4sbom/parser.py:
```diff
@@ -37,9 +37,31 @@ def __init__(self, sbom_type: str = "auto"):
+                """
+        self._parse_sbom(filename)
+
+    def parse_string(self, filename: str, file_str: str) -> None:
```
Now filename is not needed anymore

anthonyharrison reviewed (changes requested):
Let's just keep it very simple. Add a new routine parse_string which tales a JSON string to process with optional parameter to indicate type of SBOM.

trueserjo commented:
I think I already changed that. Now it's just parse_string(sbom_string) and parse_file(filename). Maybe you are looking at older changes?
As for optional parameter for type of SBOM... well, it's a bit more tricky. Let's recall first how it works. User may pass SBOM type (auto, cyclonedx or spdx) when initializing SBOMParser. If 'auto' (default) then it will try to use spdx first and fallback to cyclone on fail.
But inside spdx and cyclone parsers it analyzes filename to chose what method to use: for json, rdf, xml, jaml, or tag. You have different methods for them. E.g. if the filename ends with 'bom.json' then it's cyclone json format. So filename is needed here. That's why I wanted to keep it in parse_string.
You asked to do something to remove filename from parse_string. There are two options how to do that:
1. Ask user provide a SBOM type as a required parameter. Something like parse_string(sbom_string, sbom_type). Not just spdx or cyclone, but spdx_xml, spdx_rdf, spdx_xml, etc. I don't think it's a good option.
2. Determine sbom type (xml, rdx, xml, etc) from sbom string. I already implemeneted that. Please look at the latest changes. Now everything is simple. Just parse_string(sbom_string) - with one parameter. It took some effort and changes.
Am I missing something?

anthonyharrison commented:
@trueserjo  Nearly there. Can yopu resolve the merge conflict?

trueserjo commented:
Done.
I also added tests in test_parser.py. But one test file fails: test/data/testapp3.json. It seems that metadata -> component is a list, but parse_cyclonedx_json expects dict. I didn't change that part. It may be a bug in test file, because according to cyclonedx 1.4 json schema it should be an object.
```
"metadata": {
    "timestamp": "2023-01-24T12:57:29Z",
    "tools": [
      {
        "name": "lib4sbom",
        "version": "0.1.0"
      }
    ],
    "component": [
      {
        "type": "application",
        "bom-ref": "CDXRef-DOCUMENT",
        "name": "TestDocument"
      }
    ]
  },
```

anthonyharrison commented:
> Sure. Done. I also added tests in test_parser.py. But one test file fails: test/data/testapp3.json. It seems that metadata -> component is a list, but parse_cyclonedx_json expects dict. I didn't change that part. It may be a bug in the test file, because according to cyclonedx 1.4 json schema "component" should be an object.
> 
> ```
> "metadata": {
>     "timestamp": "2023-01-24T12:57:29Z",
>     "tools": [
>       {
>         "name": "lib4sbom",
>         "version": "0.1.0"
>       }
>     ],
>     "component": [
>       {
>         "type": "application",
>         "bom-ref": "CDXRef-DOCUMENT",
>         "name": "TestDocument"
>       }
>     ]
>   },
> ```

There are changes in the metadata format between 1.4 and the current version 1.6. Possible errors on my part. Will look at this after merge.

anthonyharrison commented:
@trueserjo The testapp3.json file was invalid. The serialNumber was incorrect and the component definition in the meatdata was wrong.

Updated file.

trueserjo commented:
@anthonyharrison Thanks!
